roachprod: refactor IAP authentication

golgeek opened the pull request:
Previously, the authentication mechanism for testeng Identity-Aware Proxy protected endpoints was relying on a shared service account key accessed via an encrypted GCS bucket. This was causing the need for secret rotation and was limiting auditability as everyone was using the same JSON key that was cached in the `~/.roachprod` directory.

This patch introduces a new mechanism based on short-lived OAuth tokens and service account impersonation through the default local credentials.

The identity of the caller is determined with the following precedence:
1. `GOOGLE_EPHEMERAL_CREDENTIALS` environment variable
2. Application Default Credentials (ADC):
  a) `GOOGLE_APPLICATION_CREDENTIALS` environment variable
  b) Default service account (application_default_credentials.json) file
  c) App Engine standard environment
  d) GCE metadata server
4. `gcloud config config-helper` output

The caller needs to have the `roles/iam.serviceAccountTokenCreator` role on the service account to be able to impersonate the service account and generate short lived OAuth AccessTokens.

Both `promhelperclient` and `grafana annotations` switch to this new method, via the new `IAPTokenSourceIface` interface that handles the service account impersonation, the AccessToken caching and renewal and that provides a pre-authenticated `http.Client`.


@cockroachdb/dev-inf, the added dependency (`github.com/binxio/gcloudconfig`) is a helper that runs `gcloud config config-helper` to get credentials from the account logged in in gcloud without the need for application default credentials file (when someone ran `gcloud auth login`, but not `gcloud auth application-default login`).

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141454)
<!-- Reviewable:end -->

golgeek commented:
Tested with local roachprod, roachprod on GCE worker, and made sure TC still has access:
- [GCE](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/18857494?buildTab=log&focusLine=3649&logView=linear&linesState=2278)
- [Azure](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAzureBazel/18857558?buildTab=log&focusLine=3726&logView=linear&linesState=2306)

Current config is: https://github.com/cockroachlabs/crl-infrastructure/blob/cbbf696fc13928c793e07d6b445cab94a26a183b/terraform/gcp/cockroachlabs.com/engineering/test-engineering/cockroach-testeng-infra/service_prometheus.tf#L13-L17

DarrylWong commented on pkg/roachprod/promhelperclient/client.go:
```diff
@@ -269,30 +245,59 @@ func buildCreateRequest(nodes map[int][]*NodeInfo, insecure bool) (io.Reader, er
+	CustomLabels map[string]string // Custom labels to be added to the cluster config
+}
+
+type Option interface {
```
Defining structs that implement the `Option` interface for every option seems unnecessary to me.

Something like: 

```
type Option interface {
	apply(*PromClient)
}

type OptionFunc func(c *PromClient)

func (o OptionFunc) apply(c *PromClient) { o(c) }

func WithIAPTokenSource(tokenSource roachprodutil.IAPTokenSourceIface) OptionFunc {
	return func(c *PromClient) {
		c.client = tokenSource.GetHTTPClient()
	}
}

func WithCustomURL(url string) OptionFunc {
	return func(c *PromClient) {
		c.promUrl = url
		c.disabled = false
	}
}
```

Should also work?

golgeek replied:
Right, much better! Thanks!

DarrylWong reviewed (commented):
>the added dependency (github.com/binxio/gcloudconfig) is a helper that runs gcloud config config-helper to get credentials from the account logged in in gcloud without the need for application default credentials file (when someone ran gcloud auth login, but not gcloud auth application-default login).

Maybe I don't understand your comment fully but I still had to rerun `gcloud auth application-default login` on my gceworker for auth to work again. Without, it was giving me 403 permission denied errors.

Lemme know if I'm missing something but if it's working as expected it might be nice to give a friendly ping in the test-eng slack that people have to reauth.  

----

Besides that, overall looks great! Just left a few nits.

golgeek commented:
> Maybe I don't understand your comment fully but I still had to rerun `gcloud auth application-default login` on my gceworker for auth to work again. Without, it was giving me 403 permission denied errors.

I'm a bit surprised because this specific change shouldn't really impact gceworkers as there will always be a service account available from the GCE metadata server, which means the order of precedence is the following:
1. env `GOOGLE_EPHEMERAL_CREDENTIALS` (if found in env)
2. env `GOOGLE_APPLICATION_CREDENTIALS` (if found in env)
3. file `APP_DATA/application_default_credentials.json` (if path exists)
4. instance's service account

Maybe I hadn't granted the gceworkers default service account authorization to impersonate the service account when you tested...

Anyway, I ran a few tests, see below:

- Not logged in on gcloud, default service account is used, success:
```bash
$ gcloud auth list
                  Credentialed Accounts
ACTIVE  ACCOUNT
*       984901724939-compute@developer.gserviceaccount.com

$ ./bin/roachprod update-targets ludoleroux-test
2025/03/18 17:30:36 client.go:140: invoking PUT for URL: https://grafana.testeng.crdb.io/promhelpers/v1/instance-configs/ludoleroux-test
```

- after `gcloud auth application-default login` with my Google account:
```bash
$ gcloud auth application-default
[...]
Credentials saved to file: [/home/ludo_leroux_cockroachlabs_com/.config/gcloud/application_default_credentials.json]

$ ./bin/roachprod update-targets ludoleroux-test
2025/03/18 17:33:29 client.go:140: invoking PUT for URL: https://grafana.testeng.crdb.io/promhelpers/v1/instance-configs/ludoleroux-test
```

---

The change in behavior that this additional package brings happens outside of GCP when there is no environment variable or application default credentials, but user is logged-in through `gcloud`:
```bash
$ env | grep GOOGLE
$ gcloud auth application-default revoke
You are about to revoke the credentials stored in: [/Users/ludovicleroux/.config/gcloud/application_default_credentials.json]

Do you want to continue (Y/n)?  Y

Credentials revoked.

$ gcloud auth list
        Credentialed Accounts
ACTIVE  ACCOUNT
*       ludo.leroux@cockroachlabs.com
$ ./bin/roachprod update-targets ludoleroux-test
2025/03/18 17:50:24 client.go:140: invoking PUT for URL: https://grafana.testeng.crdb.io/promhelpers/v1/instance-configs/ludoleroux-test
```

golgeek commented on pkg/roachprod/promhelperclient/client.go:
```diff
@@ -237,20 +178,55 @@ func ProviderReachability(provider string, cloudEnvironment CloudEnvironment) Re
+	if err != nil {
+		return errors.Wrapf(err, "%s: failed on url: %s", logPrefix, request.URL)
+	}
+	if response.StatusCode != expectedStatus {
```
Yeah, I completely got tripped by the existing logic (which I think was mostly around for the secure/insecure retry logic that doesn't exist anymore)

DarrylWong commented on pkg/roachprod/promhelperclient/client.go:
```diff
@@ -237,20 +178,53 @@ func ProviderReachability(provider string, cloudEnvironment CloudEnvironment) Re
+func (c *PromClient) makeRequest(request *http.Request, logPrefix string) error {
+	response, err := c.client.Do(request)
+	if err != nil {
+		return errors.Wrapf(err, "%s: failed on url: %s", logPrefix, request.URL)
```
nit: it probably makes more sense for the caller of `makeRequest` to do the error wrapping - if they want

DarrylWong commented on pkg/roachprod/promhelperclient/client.go:
```diff
@@ -91,131 +93,70 @@ type PromClient struct {
-func IsNotFoundError(err error) bool {
-	return strings.Contains(err.Error(), fmt.Sprintf(ErrorMessagePrefix, http.StatusNotFound))
+	// client is the http client
+	client *http.Client
```
nit: why not `httpClient` to remove any ambiguity?

DarrylWong commented on pkg/roachprod/roachprodutil/identity.go:
```diff
@@ -7,129 +7,129 @@ package roachprodutil
-type FetchedFrom string
+// IAPTokenSourceIface is an interface that defines the methods
+// for the IAPTokenSource.
+type IAPTokenSourceIface interface {
```
nit: this being `IAPTokenSource` and the implementation below being `IAPTokenSourceImpl` is more idiomatic with the rest of the roachtest/roachprod code

golgeek committed:
roachprod: refactor IAP authentication

Previously, the authentication mechanism for testeng Identity-Aware
Proxy protected endpoints was relying on a shared ervice account key
accessed via a GCS bucket. This was causing the need for secret rotation
and was limiting auditability.

This patch introduces a new mechanism based on short-lived OAuth tokens
and service account impersonation through the default local credentials.

The identity of the caller is determined with the following precedence:
1. `GOOGLE_EPHEMERAL_CREDENTIALS` environment variable
2. Application Default Credentials (ADC):
  a. `GOOGLE_APPLICATION_CREDENTIALS` environment variable
  b. Default service account (application_default_credentials.json) file
  c. App Engine standard environment
  d. GCE metadata server
3. `gcloud config config-helper` output

The caller needs to have the `roles/iam.serviceAccountTokenCreator` role
on the service account to be able to impersonate the service account and
generate short lived OAuth AccessTokens.

Both `promhelperclient` and `grafana annotations` switch to this new
method, via the new `IAPTokenSourceIface` interface that handles the
service account impersonation, the AccessToken caching and renewal and
that provides a pre-authenticated `http.Client`.

Epic: none
Release note: None
